roachtest: sqlsmith/setup=rand-tables/setting=no-mutations failed

[(roachtest).sqlsmith/setup=rand-tables/setting=no-mutations failed](https://teamcity.cockroachdb.com/viewLog.html?buildId=2819724&tab=buildLog) on [master@3cfe2a38044b9e0d47b09815658e8634e4f4bfda](https://github.com/cockroachdb/cockroach/commits/3cfe2a38044b9e0d47b09815658e8634e4f4bfda):

```
The test failed on branch=master, cloud=gce:
test artifacts and logs in: /home/agent/work/.go/src/github.com/cockroachdb/cockroach/artifacts/sqlsmith/setup=rand-tables/setting=no-mutations/run_1
	sqlsmith.go:226,sqlsmith.go:262,test_runner.go:768: error: pq: internal error: parameter unknown is not float
		stmt:
		SELECT
			count(*) AS col_9954
		FROM
			defaultdb.public.table3@[0] AS tab_4368
		GROUP BY
			tab_4368.col3_2
		HAVING
			st_dfullywithin(tab_4368.col3_2::GEOMETRY, '0103000080010000000D000000B69EA24CD242F141F4A559F67BA5F6C19A2F8FA3DC7AF0419C8E7784A933F3414E0FF2D93E17E7C156511FE87EE5FCC11CB45A4E3ED2E741C0942722B7F7D0C10054C7671E8E9A41D4797FCB1E41F741A06E052316EBDDC11C8098700DBBF84182BA05798ACEF74100DA077053687E415E0F2838028602429C3B88DFBA48EA416045DA4F1C3CBC41F410037434C9FF419E4A16022C9AFA4120989642CE17C041009F2570447AD04174818A4626F8FAC158AB9F50C7F6004240EF0371DF6ECB414F98404580B2FAC1C4E66AE3689BF24152D18747B883E4C19E53C0A6B3E8FFC1B44C1A3BA069F44168CAA9D5010401424C56C6F8754EF9C11C81A971ECEEE3417707383A674EF5C1F36F3081ED0302C2B03092CBF89BD041A416AA1CC7B4FBC1B69EA24CD242F141F4A559F67BA5F6C19A2F8FA3DC7AF041':::GEOMETRY::GEOMETRY, NULL::FLOAT8)::BOOL
		ORDER BY
			tab_4368.col3_2, tab_4368.col3_2 ASC
		LIMIT
			78:::INT8;

```

<details><summary>More</summary><p>


Artifacts: [/sqlsmith/setup=rand-tables/setting=no-mutations](https://teamcity.cockroachdb.com/viewLog.html?buildId=2819724&tab=artifacts#/sqlsmith/setup=rand-tables/setting=no-mutations)

[See this test on roachdash](https://roachdash.crdb.dev/?filter=status%3Aopen+t%3A.%2Asqlsmith%2Fsetup%3Drand-tables%2Fsetting%3Dno-mutations.%2A&sort=title&restgroup=false&display=lastcommented+project)
<sub>powered by [pkg/cmd/internal/issues](https://github.com/cockroachdb/cockroach/tree/master/pkg/cmd/internal/issues)</sub></p></details>

--- a/pkg/sql/logictest/testdata/logic_test/inverted_join_geospatial
+++ b/pkg/sql/logictest/testdata/logic_test/inverted_join_geospatial
@@ -477,3 +477,15 @@ FULL OUTER JOIN
 ON inv_join.k1 = cross_join.k1 AND inv_join.k2 = cross_join.k2
 WHERE inv_join.k1 IS NULL OR cross_join.k1 IS NULL
 ----
+
+# Regression test for #62686. An inverted join with a geospatial function and a
+# NULL distance argument should not error.
+statement ok
+CREATE TABLE t62686 (
+  c GEOMETRY,
+  INVERTED INDEX (c ASC)
+);
+INSERT INTO t62686 VALUES (ST_GeomFromText('POINT(1 1)'));
+
+statement ok
+SELECT * FROM t62686 t1 JOIN t62686 t2 ON ST_DFullyWithin(t1.c, t2.c, NULL::FLOAT8)

--- a/pkg/sql/opt/invertedidx/geo.go
+++ b/pkg/sql/opt/invertedidx/geo.go
@@ -139,8 +139,9 @@ func getSpanExprForGeographyIndex(
 
 // Helper for DWithin and DFullyWithin.
 func getDistanceParam(params []tree.Datum) float64 {
-	// Parameters are type checked earlier. Keep this consistent with the definition
-	// in geo_builtins.go.
+	// Parameters are type checked earlier when the expression is built by
+	// optbuilder. extractInfoFromExpr ensures that the parameters are non-NULL
+	// constants. Keep this consistent with the definition in geo_builtins.go.
 	if len(params) != 1 {
 		panic(errors.AssertionFailedf("unexpected param length %d", len(params)))
 	}
@@ -456,6 +457,11 @@ func extractInfoFromExpr(
 		arg1, exprArg2 = exprArg2, arg1
 	}
 
+	// The first argument must be non-NULL.
+	if arg1.Op() == opt.NullOp {
+		return 0, nil, nil, nil, false
+	}
+
 	// The second argument should be a variable corresponding to the index
 	// column.
 	arg2, ok = exprArg2.(*memo.VariableExpr)
@@ -467,9 +473,10 @@ func extractInfoFromExpr(
 		return 0, nil, nil, nil, false
 	}
 
-	// Any additional params must be constant.
+	// Any additional params must be non-NULL constants.
 	for i := 2; i < args.ChildCount(); i++ {
-		if !memo.CanExtractConstDatum(args.Child(i)) {
+		arg := args.Child(i)
+		if arg.Op() == opt.NullOp || !memo.CanExtractConstDatum(arg) {
 			return 0, nil, nil, nil, false
 		}
 		additionalParams = append(additionalParams, memo.ExtractConstDatum(args.Child(i)))

--- a/pkg/sql/opt/xform/testdata/rules/select
+++ b/pkg/sql/opt/xform/testdata/rules/select
@@ -4823,6 +4823,53 @@ select
  └── filters
       └── st_intersects(g:2, '0103000020E610000000000000') [outer=(2), immutable, constraints=(/2: (/NULL - ])]
 
+# Regression test for #60527. Do not panic when a geospatial function has a NULL
+# GEOMETRY/GEOGRAPHY argument.
+exec-ddl
+CREATE TABLE t60527 (
+  g GEOGRAPHY,
+  INVERTED INDEX (g)
+)
+----
+
+# TODO(mgartner): Functions with NullableArgs=false and a NULL argument can be
+# normalized to NULL. This would simplify this query plan to an empty Values
+# expression.
+opt
+SELECT * FROM t60527 WHERE (ST_Covers(NULL::GEOGRAPHY, g) AND ST_DWithin(NULL::GEOGRAPHY, g, 1))
+----
+select
+ ├── columns: g:1!null
+ ├── immutable
+ ├── scan t60527
+ │    └── columns: g:1
+ └── filters
+      ├── st_covers(CAST(NULL AS GEOGRAPHY), g:1) [outer=(1), immutable, constraints=(/1: (/NULL - ])]
+      └── st_dwithin(CAST(NULL AS GEOGRAPHY), g:1, 1.0) [outer=(1), immutable, constraints=(/1: (/NULL - ])]
+
+# Regression test for #62686. Do not panic when a geospatial function has a NULL
+# additional argument.
+exec-ddl
+CREATE TABLE t62686 (
+  c GEOMETRY NULL,
+  INVERTED INDEX (c ASC)
+)
+----
+
+# TODO(mgartner): Functions with NullableArgs=false and a NULL argument can be
+# normalized to NULL. This would simplify this query plan to an empty Values
+# expression.
+opt
+SELECT * FROM t62686 WHERE ST_DFullyWithin(c, ST_GeomFromText('POINT(1 1)'), NULL::FLOAT8)
+----
+select
+ ├── columns: c:1!null
+ ├── immutable
+ ├── scan t62686
+ │    └── columns: c:1
+ └── filters
+      └── st_dfullywithin(c:1, '0101000000000000000000F03F000000000000F03F', CAST(NULL AS FLOAT8)) [outer=(1), immutable, constraints=(/1: (/NULL - ])]
+
 # --------------------------------------------------
 # GenerateZigzagJoins
 # --------------------------------------------------

